Lib/dataclasses.py

```diff
@@ -1318,23 +1352,33 @@ def _asdict_inner(obj, dict_factory):
         #   namedtuples, we could no longer call asdict() on a data
         #   structure where a namedtuple was used as a dict key.
 
-        return type(obj)(*[_asdict_inner(v, dict_factory) for v in obj])
+        return type(obj)(*[
+            v if type(v) in _ATOMIC_TYPES else _asdict_inner(v, dict_factory)
+            for v in obj
+        ])
```

Comment:
I find it so interesting that avoiding the recursive call here appears to save us so much time. Initially when I looked at this PR I wondered if we could do something slightly simpler (https://github.com/python/cpython/compare/main...AlexWaygood:cpython:asdict-faster), but it seems like that is substantially slower on some of your benchmarks!

Reply:
I'm using a non-debug, PGO-optimised build already -- I'm learning! ;)

Here's a three-way comparison on my machine, using @DavidCEllis's benchmark (which looks reasonable to me, but I'm no expert on writing benchmarks):

<details>
<summary>`main` branch</summary>

```
Dataclasses asdict/astuple speed tests
--------------------------------------
Python v3.12.0alpha6
GIT branch: main
Test Iterations: 10000
List of Int case asdict: 5.71s

Test Iterations: 1000
List of Decimal case asdict: 0.63s

Test Iterations: 1000000
Basic types case asdict: 9.35s
Basic types astuple: 11.84s

Test Iterations: 100000
Opaque types asdict: 7.27s
Opaque types astuple: 7.06s

Test Iterations: 100
Mixed containers asdict: 11.63s
Mixed containers astuple: 10.74s
```

</details>

<details>
<summary>This PR</summary>

```
Running PGUpdate|x64 interpreter...
Dataclasses asdict/astuple speed tests
--------------------------------------
Python v3.12.0alpha6
GIT branch: faster_dataclasses_serialize
Test Iterations: 10000
List of Int case asdict: 0.60s

Test Iterations: 1000
List of Decimal case asdict: 0.65s

Test Iterations: 1000000
Basic types case asdict: 1.29s
Basic types astuple: 1.26s

Test Iterations: 100000
Opaque types asdict: 2.00s
Opaque types astuple: 2.15s

Test Iterations: 100
Mixed containers asdict: 6.48s
Mixed containers astuple: 6.37s
```

</details>

<details>
<summary>Alex's "simpler" version</summary>

```
Running PGUpdate|x64 interpreter...
Dataclasses asdict/astuple speed tests
--------------------------------------
Python v3.12.0alpha6
GIT branch: asdict-faster
Test Iterations: 10000
List of Int case asdict: 1.31s

Test Iterations: 1000
List of Decimal case asdict: 0.67s

Test Iterations: 1000000
Basic types case asdict: 1.76s
Basic types astuple: 1.67s

Test Iterations: 100000
Opaque types asdict: 2.02s
Opaque types astuple: 5.93s

Test Iterations: 100
Mixed containers asdict: 7.74s
Mixed containers astuple: 6.88s
```

</details>

So: it does seem like both this PR and my "simpler" version would be substantially faster than `main`. Also: I believe Python-to-Python calls are meant to be exactly the kind of thing that the faster-CPython team are trying to optimise in the interpreter core itself? So maybe we shouldn't be trying to micro-optimise based on the slowness of Python-to-Python calls _too_ much.

Reply:
Thanks for doing that benchmarking!

I agree with you that we shouldn't overly micro-optimize based on function calls being slow: it's the responsibility of interpreter optimizations to reduce that cost so we don't have to contort the Python code we write. And the simpler version where we bail out early at the top of `_as*_inner` for atomic types still offers the bulk of the performance win, with much less repetitive code. So I think we should go for the simpler version here, and not sprinkle identical checks against `_ATOMIC_TYPES` throughout every case in these functions.

Reply:
Tbh I'm also curious about the split of the win between the `default_factory` optimization and the `_ATOMIC_TYPES` one. I assume both are significant, and the cost of the `default_factory` change in code complexity is also pretty low, but it would be ideal to evaluate the cost/benefit of each independently.

Reply:
> I assume both are significant

Hmm, I tried benchmarking, and I'm not sure the `default_factory` optimisation actually gets us much. Here's benchmark results from https://github.com/python/cpython/compare/main...AlexWaygood:cpython:asdict-faster-no-default-factory-change, which is my "simple" version but with the `default_factory` change removed. It's hard to see much difference between these results and the results from my "simple" version including the `default_factory` change:

<details>

```
Running PGUpdate|x64 interpreter...
Dataclasses asdict/astuple speed tests
--------------------------------------
Python v3.12.0alpha6
GIT branch: asdict-faster-no-default-factory-change
Test Iterations: 10000
List of Int case asdict: 1.16s

Test Iterations: 1000
List of Decimal case asdict: 0.67s

Test Iterations: 1000000
Basic types case asdict: 1.87s
Basic types astuple: 1.55s

Test Iterations: 100000
Opaque types asdict: 2.04s
Opaque types astuple: 5.64s

Test Iterations: 100
Mixed containers asdict: 7.94s
Mixed containers astuple: 6.90s
```

</details>

So: +1 to considering this optimisation separately, and maybe moving it to a different PR.

Reply:
> The `default_factory` optimization is much smaller than the `_ATOMIC_TYPES` one, but was still enough to be consistent and measurable. I can add a performance test that uses a plain subclass for `default_factory`.

Sure, but let's consider each change in isolation: it's a distinct change that should be a separate PR.

Reply:
Updated: https://gist.github.com/DavidCEllis/a2c2ceeeeda2d1ac509fb8877e5fb60d

On some tests it's not enough of the overhead to matter, but it's fairly clear on some of the others (for me at least).

<details>
<summary>Perf result</summary>

```
Dataclasses asdict/astuple speed tests
--------------------------------------
Python v3.12.0alpha6
GIT branch: faster_dataclasses_serialize
Test Iterations: 10000 | dict_factory=dict
List of Int case asdict: 0.55s

Test Iterations: 10000 | dict_factory=CustomDict
List of Int case asdict: 0.56s

Test Iterations: 1000 | dict_factory=dict
List of Decimal case asdict: 0.73s

Test Iterations: 1000 | dict_factory=CustomDict
List of Decimal case asdict: 0.74s

Test Iterations: 1000000 | dict_factory=dict
Basic types case asdict: 1.37s
Basic types astuple: 1.29s

Test Iterations: 1000000 | dict_factory=CustomDict
Basic types case asdict: 1.67s

Test Iterations: 100000 | dict_factory=dict
Opaque types asdict: 2.32s
Opaque types astuple: 2.29s

Test Iterations: 100000 | dict_factory=CustomDict
Opaque types asdict: 2.35s

Test Iterations: 100 | dict_factory=dict
Mixed containers asdict: 2.08s
Mixed containers astuple: 2.00s

Test Iterations: 100 | dict_factory=CustomDict
Mixed containers asdict: 2.41s

```
</details>

Reply:
Great! However, adding new benchmarks does not address my argument that the `default_factory` change is a separate optimisation that should be evaluated independently as part of a separate PR

Reply:
Sorry, the problem is mostly that due to the interconnected changes I couldn't really offer it as 2 separate PRs without first having 1 accepted. I guess I can remove this for now.

Reply:
You can attach the followup PR to the same issue (you can probably share one news entry between them :)

Reply:
I've removed it for now. I think this threading system in github is going to give me a headache though!